mlflow/deployments/base.py

```diff
@@ -70,6 +70,208 @@ def target_help():
     )
 
 
+class BaseEndpointClient(abc.ABC):
+    """
+    Base class exposing Python model endpoint APIs.
+
+    Plugin implementers should define target-specific logic via a subclass of
+    ``BaseEndpointClient`` within the plugin module, and customize method docstrings with
+    target-specific information.
+
+    .. Note::
+        Subclasses should raise :py:class:`mlflow.exceptions.MlflowException` in error cases
+        (e.g. on failure to deploy a model).
+    """
+
+    def __init__(self):
+        pass
```

Comment:
@jinzhang21 I saw that we removed `target_uri` based on feedback from https://github.com/mlflow/mlflow/pull/5378#discussion_r842247310. `target_uri` is used to specify where deployment resources should be created - e.g. a particular AML workspace, AWS account, etc. Without such a capability, everything needs to be passed via `config`. I think it should be included. This gets propagated down in the deployment client case via https://mlflow.org/docs/latest/python_api/mlflow.deployments.html#mlflow.deployments.get_deploy_client.

@trangevi Independently of this, how do we envision users interacting with implementations of `BaseEndpointClient`? If I want to interact with a deployment client via the Python API, I run something like the following:

```
client = mlflow.deployments.get_deploy_client("<azureml_tracking_uri>")
client.create_deployment(...)
```

Will users do something like the following for an endpoint client?

```
endpoint_client = mlflow.deployments.get_endpoint_client("<azureml_tracking_uri>")
endpoint_client.create_endpoint(...)
```

To me, exposing these methods on the user-facing deployment client seems more natural:

```
client = mlflow.deployments.get_deploy_client("<azureml_tracking_uri>")
client.create_deployment(...)
client.create_endpoint(...)
```

Is this the intention? If so, should `BaseEndpointClient` be a distinct abstract class? I think that's fine as long as there's a plan for exposing `BaseEndpointClient` methods in the user-facing result of `mlflow.deployments.get_endpoint_client`, but we should make sure to align on the UX and implementation details for this before moving forward since it isn't obvious.

Reply:
@dbczumar, the discussion that we'd had to add a distinct BaseEndpointClient was based on not ever plugin implementer having an endpoints concept. This was decided as the best way to support both scenarios while being clearest to users of the clients what they can interact with (ie not having deployment methods with optional endpoint parameters that may or may not make sense).

I can add a `get_endpoint_client` method to match the deployment one, if that's the missing piece for exposing it as a distinct entity. And add back target_uri to match.

Reply:
On adding back target_uri, can we add comments on why it's needed and how it will be used? It wasn't clear in the initial commit, leading to a suggested removal.

Reply:
@trangevi I think we need to consider / align on the user experience before merging this code. As it stands, the code introduces a developer-facing `BaseEndpointClient` without creating any corresponding user-facing APIs for interacting with it. I'm also hesitant to bifurcate the user-facing API surface into deployment APIs exposed by `DeploymentClient` and separate endpoint APIs exposed by `EndpointClient`, since they two are conceptually intertwined. I'd favor the following UX:

```
client = mlflow.deployments.get_deploy_client("<azureml_tracking_uri>")
client.create_deployment(...)
client.create_endpoint(...)
```

If this means that we have to merge `BaseEndpointClient` with `BaseDeploymentClient`, that's okay; instead of marking these methods as explicitly abstract via `abc.abstractmethod`, we can have them raise `NotImplemented` and include informative docstrings explaining their intended use cases and telling developers when they should be implemented.

Reply:
Discussed offline and made corresponding changes